File: src/main/java/com/aquasecurity/plugins/trivy/settings/TrivySettingsConfigurable.java
package com.aquasecurity.plugins.trivy.settings;


import com.intellij.openapi.options.Configurable;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

/**
 * Provides controller functionality for application settings.
 */
public class TrivySettingsConfigurable implements Configurable {

    private TrivySettingsComponent TrivySettingsComponent;

    // A default constructor with no arguments is required because this implementation
    // is registered as an applicationConfigurable EP

    @Nls(capitalization = Nls.Capitalization.Title)
    @Override
    public String getDisplayName() {
        return "Trivy: Settings";
    }

    @Override
    public JComponent getPreferredFocusedComponent() {
        return TrivySettingsComponent.getPreferredFocusedComponent();
    }

    @Nullable
    @Override
    public JComponent createComponent() {
        TrivySettingsComponent = new TrivySettingsComponent();
        return TrivySettingsComponent.getPanel();
    }

    @Override
    public boolean isModified() {
        TrivySettingState settings = TrivySettingState.getInstance();
        boolean modified = !TrivySettingsComponent.getTrivyPath().equals(settings.TrivyPath) ||
                !TrivySettingsComponent.getCriticalSeverityRequired() == settings.CriticalSeverity ||
                !TrivySettingsComponent.getHighSeverityRequired() == settings.HighSeverity ||
                !TrivySettingsComponent.getMediumSeverityRequired() == settings.MediumSeverity||
                !TrivySettingsComponent.getLowSeverityRequired() == settings.LowSeverity         ||
                !TrivySettingsComponent.getUnknownSeverityRequired() == settings.UnknownSeverity ||
                !TrivySettingsComponent.getOfflineScanRequired() == settings.OfflineScan ||
                !TrivySettingsComponent.getShowOnlyFixed() == settings.IgnoreUnfixed
                ;
        return modified;
    }

    @Override
    public void apply() {
        TrivySettingState settings = TrivySettingState.getInstance();
        settings.TrivyPath = TrivySettingsComponent.getTrivyPath();
        settings.CriticalSeverity = TrivySettingsComponent.getCriticalSeverityRequired();
        settings.HighSeverity = TrivySettingsComponent.getHighSeverityRequired();
        settings.MediumSeverity = TrivySettingsComponent.getMediumSeverityRequired();
        settings.LowSeverity = TrivySettingsComponent.getLowSeverityRequired();
        settings.UnknownSeverity = TrivySettingsComponent.getUnknownSeverityRequired();
        settings.OfflineScan = TrivySettingsComponent.getOfflineScanRequired();
        settings.IgnoreUnfixed = TrivySettingsComponent.getShowOnlyFixed();

    }

    @Override
    public void reset() {
        TrivySettingState settings = TrivySettingState.getInstance();
        TrivySettingsComponent.setTrivyPath(settings.TrivyPath);
        TrivySettingsComponent.setCriticalSeverity(settings.CriticalSeverity);
        TrivySettingsComponent.setHighSeverity(settings.HighSeverity);
        TrivySettingsComponent.setMediumSeverity(settings.MediumSeverity);
        TrivySettingsComponent.setLowSeverity(settings.LowSeverity);
        TrivySettingsComponent.setUnknownSeverity(settings.UnknownSeverity);
        TrivySettingsComponent.setOfflineScan(settings.OfflineScan);
        TrivySettingsComponent.setIgnoreUnfixed(settings.IgnoreUnfixed);

    }

    @Override
    public void disposeUIResources() {
        TrivySettingsComponent = null;
    }

}

File: src/main/kotlin/com/aquasecurity/plugins/trivy/settings/TrivySettingsConfigurable.kt
package com.aquasecurity.plugins.trivy.settings

import com.aquasecurity.plugins.trivy.actions.CheckForTrivyAction
import com.intellij.openapi.options.Configurable
import com.intellij.openapi.project.Project
import javax.swing.JComponent
import org.jetbrains.annotations.Nls

/** Provides controller functionality for application settings. */
class TrivySettingsConfigurable(private val project: Project) : Configurable {
  private var trivySettingsComponent: TrivySettingsComponent? = null

  // A default constructor with no arguments is required because this implementation
  // is registered as an applicationConfigurable EP
  override fun getDisplayName(): @Nls(capitalization = Nls.Capitalization.Title) String {
    return "Trivy: Settings"
  }

  override fun getPreferredFocusedComponent(): JComponent? {
    return trivySettingsComponent?.preferredFocusedComponent
  }

  override fun createComponent(): JComponent? {
    trivySettingsComponent = TrivySettingsComponent()
    return trivySettingsComponent?.panel
  }

  override fun isModified(): Boolean {
    val settings = TrivySettingState.instance
    var projectSettings = TrivyProjectSettingState.getInstance(project)

    if (trivySettingsComponent == null) {
      return false
    }

    val modified =
        (trivySettingsComponent!!.getTrivyPath() != settings.trivyPath ||
            !trivySettingsComponent!!.getCriticalSeverityRequired == settings.criticalSeverity ||
            !trivySettingsComponent!!.getHighSeverityRequired == settings.highSeverity ||
            !trivySettingsComponent!!.getMediumSeverityRequired == settings.mediumSeverity ||
            !trivySettingsComponent!!.getLowSeverityRequired == settings.lowSeverity ||
            !trivySettingsComponent!!.getUnknownSeverityRequired == settings.unknownSeverity ||
            !trivySettingsComponent!!.getOfflineScanRequired == settings.offlineScan ||
            !trivySettingsComponent!!.getShowOnlyFixed == settings.ignoreUnfixed ||
            !trivySettingsComponent!!.getScanForSecrets == settings.scanForSecrets ||
            !trivySettingsComponent!!.getScanForMisconfigurations ==
                settings.scanForMisconfigurations ||
            !trivySettingsComponent!!.getScanForVulnerabilities ==
                settings.scanForVulnerabilities ||
            trivySettingsComponent!!.getConfigPath() != projectSettings.configPath ||
            !trivySettingsComponent!!.getUseConfig == projectSettings.useConfig ||
            trivySettingsComponent!!.getIgnorePath() != projectSettings.ignorePath ||
            !trivySettingsComponent!!.getUseIgnore == projectSettings.useIgnore ||
            trivySettingsComponent!!.getApiKey != settings.apiKey ||
            trivySettingsComponent!!.getApiSecret != settings.apiSecret ||
            trivySettingsComponent!!.getRegion != settings.region ||
            trivySettingsComponent!!.getUseAquaPlatform != projectSettings.useAquaPlatform ||
            trivySettingsComponent!!.getEnableDotNetProject !=
                projectSettings.enableDotNetProject ||
            trivySettingsComponent!!.getEnableGradle != projectSettings.enableGradle ||
            trivySettingsComponent!!.getEnablePackageJson != projectSettings.enablePackageJson ||
            trivySettingsComponent!!.getEnableSASTScanning != projectSettings.enableSASTScanning ||
            trivySettingsComponent!!.getCustomAquaUrl != settings.customAquaUrl ||
            trivySettingsComponent!!.getCustomAuthUrl != settings.customAuthUrl)

    return modified
  }

  override fun apply() {
    val settings = TrivySettingState.instance
    val projectSettings = TrivyProjectSettingState.getInstance(project)
    if (trivySettingsComponent == null) {
      return
    }

    settings.trivyPath = trivySettingsComponent!!.getTrivyPath()
    settings.criticalSeverity = trivySettingsComponent!!.getCriticalSeverityRequired
    settings.highSeverity = trivySettingsComponent!!.getHighSeverityRequired
    settings.mediumSeverity = trivySettingsComponent!!.getMediumSeverityRequired
    settings.lowSeverity = trivySettingsComponent!!.getLowSeverityRequired
    settings.unknownSeverity = trivySettingsComponent!!.getUnknownSeverityRequired
    settings.offlineScan = trivySettingsComponent!!.getOfflineScanRequired
    settings.ignoreUnfixed = trivySettingsComponent!!.getShowOnlyFixed
    settings.scanForSecrets = trivySettingsComponent!!.getScanForSecrets
    settings.scanForMisconfigurations = trivySettingsComponent!!.getScanForMisconfigurations
    settings.scanForVulnerabilities = trivySettingsComponent!!.getScanForVulnerabilities
    settings.apiKey = trivySettingsComponent!!.getApiKey
    settings.apiSecret = trivySettingsComponent!!.getApiSecret
    settings.region = trivySettingsComponent!!.getRegion

    if (settings.region == "Custom") {
      settings.customAquaUrl = trivySettingsComponent!!.getCustomAquaUrl
      settings.customAuthUrl = trivySettingsComponent!!.getCustomAuthUrl
    } else {
      settings.customAquaUrl = ""
      settings.customAuthUrl = ""
    }

    projectSettings.configPath = trivySettingsComponent!!.getConfigPath()
    projectSettings.useConfig = trivySettingsComponent!!.getUseConfig
    projectSettings.ignorePath = trivySettingsComponent!!.getIgnorePath()
    projectSettings.useIgnore = trivySettingsComponent!!.getUseIgnore
    projectSettings.enableDotNetProject = trivySettingsComponent!!.getEnableDotNetProject
    projectSettings.enableGradle = trivySettingsComponent!!.getEnableGradle
    projectSettings.enablePackageJson = trivySettingsComponent!!.getEnablePackageJson
    projectSettings.enableSASTScanning = trivySettingsComponent!!.getEnableSASTScanning

    if (trivySettingsComponent!!.getUseAquaPlatform) {
      projectSettings.useAquaPlatform =
          CredentialCheck.isValidCredentials(
              project,
              settings.apiKey,
              settings.apiSecret,
              settings.customAquaUrl,
              settings.customAuthUrl)
    }
    CheckForTrivyAction.run(project)
  }

  override fun reset() {
    val settings = TrivySettingState.instance
    trivySettingsComponent?.setTrivyPath(settings.trivyPath)
    trivySettingsComponent?.setCriticalSeverity(settings.criticalSeverity)
    trivySettingsComponent?.setHighSeverity(settings.highSeverity)
    trivySettingsComponent?.setMediumSeverity(settings.mediumSeverity)
    trivySettingsComponent?.setLowSeverity(settings.lowSeverity)
    trivySettingsComponent?.setUnknownSeverity(settings.unknownSeverity)
    trivySettingsComponent?.setOfflineScan(settings.offlineScan)
    trivySettingsComponent?.setIgnoreUnfixed(settings.ignoreUnfixed)
    trivySettingsComponent?.setSecretScanning(settings.scanForSecrets)
    trivySettingsComponent?.setMisconfigurationScanning(settings.scanForMisconfigurations)
    trivySettingsComponent?.setVulnerabilityScanning(settings.scanForVulnerabilities)
    trivySettingsComponent?.setConfigFilePath(
        TrivyProjectSettingState.getInstance(project).configPath)
    trivySettingsComponent?.setUseConfig(TrivyProjectSettingState.getInstance(project).useConfig)
    trivySettingsComponent?.setIgnoreFilePath(
        TrivyProjectSettingState.getInstance(project).ignorePath)
    trivySettingsComponent?.setUseIgnore(TrivyProjectSettingState.getInstance(project).useIgnore)
    trivySettingsComponent?.setApiKey(settings.apiKey)
    trivySettingsComponent?.setApiSecret(settings.apiSecret)
    trivySettingsComponent?.setRegion(settings.region)

    if (settings.customAquaUrl.isNotEmpty()) {
      trivySettingsComponent?.setCustomAquaUrl(settings.customAquaUrl)
    }
    if (settings.customAuthUrl.isNotEmpty()) {
      trivySettingsComponent?.setCustomAuthUrl(settings.customAuthUrl)
    }

    trivySettingsComponent?.setUseAquaPlatform(
        TrivyProjectSettingState.getInstance(project).useAquaPlatform)
    trivySettingsComponent?.setEnableDotNetProject(
        TrivyProjectSettingState.getInstance(project).enableDotNetProject)
    trivySettingsComponent?.setEnableGradle(
        TrivyProjectSettingState.getInstance(project).enableGradle)
    trivySettingsComponent?.setEnablePackageJson(
        TrivyProjectSettingState.getInstance(project).enablePackageJson)
    trivySettingsComponent?.setEnableSASTScanning(
        TrivyProjectSettingState.getInstance(project).enableSASTScanning)
  }

  override fun disposeUIResources() {
    trivySettingsComponent = null
  }
}
